File: 프로그래머스/lv2/홍혁재/타겟 넘버.cpp
/*
알고리즘: 브루트 포스
*/

#include <string>
#include <vector>

using namespace std;

int bf(vector<int> numbers, int target, int x, int total) {
    if(x >= numbers.size()) {
        if(total == target) return 1;
        else return 0;
    }
    for(int i = x; i < numbers.size(); i++) {
        return bf(numbers, target, i + 1, total + numbers[i]) + 
            bf(numbers, target, i + 1, total - numbers[i]);;
    }
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    
    answer = bf(numbers, target, 0, 0);
    
    return answer;
}

File: 프로그래머스/lv2/최원준/타겟 넘버.py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :
    해당 숫자를 더하거나 뺀 결과를 저장하고, 저장된 결과에서 또 더하고 뺴고를 반복
2. 시간복잡도 :
    O(n)
3. 자료구조 :
    해시맵
'''


from collections import defaultdict
def solution(numbers, target):
    prev = defaultdict(int)
    prev[0] = 1

    for i, n in enumerate(numbers):
        curr = defaultdict(int)
        for k, v in prev.items():
            curr[k+n] += v
            curr[k-n] += v
        prev = curr
    return prev[target]
